### ISSUE TITLE
Windows eventchannel decoder is not compatible with other XML decoders

### ISSUE THREAD

**Issue by cristgl** [wazuh/wazuh#3193]
|Wazuh version|Install type|Install method|Platform|
|---|---|---|---|
| 3.9.0-3916 | Manager | Packages/Sources | Windows/Linux |

This issue was inspired by the user mcarnerero at the [lists](https://groups.google.com/forum/#!topic/wazuh/XTKfGGGezYw).

Events decoded by the Eventchannel decoder can't be processed again with other decoders, as it won't get back to the decoding phase once it is processed by `analysisd`.

This could be an interesting use case where we may use the XML decoders after the Eventchannel one: 

- The next Windows event has a relevant field which name is `DeviceId`. It contains different information about the USB connected in this case.

```
{
   "EventChannel":{
      "System":{
         "ProviderName":"Microsoft-Windows-Security-Auditing",
         "ProviderGuid":"{54849625-5478-4994-A5BA-3E3B0328C30D}",
         "EventID":"6416",
         "Version":"1",
         "Level":"0",
         "Task":"13316",
         "Opcode":"0",
         "Keywords":"0x8020000000000000",
         "SystemTime":"2019-04-10T10:27:29.598582700Z",
         "EventRecordID":"2840",
         "ProcessID":"4",
         "ThreadID":"4312",
         "Channel":"Security",
         "Computer":"DESKTOP-DJ7F955",
         "SeverityValue":"AUDIT_SUCCESS",
         "Message":"A new external device was recognized by the system."
      },
      "EventData":{
         "SubjectUserSid":"S-1-5-18",
         "SubjectUserName":"DESKTOP-DJ7F955$",
         "SubjectDomainName":"WORKGROUP",
         "SubjectLogonId":"0x3e7",
         "DeviceId":"SWD\\WPDBUSENUM\\_??_USBSTOR#Disk&amp;Ven_General&amp;Prod_USB_Flash_Disk&amp;Rev_1100#0411150000016636&amp;0#{53f56307-b6bf-11d0-94f2-00a0c91efb8b}",
         "DeviceDescription":"USB MALWARE",
         "ClassId":"{EEC5AD98-8080-425F-922A-DABF3DE3F69A}",
         "ClassName":"WPD",
         "VendorIds":"-",
         "CompatibleIds":"wpdbusenum\\fsSWD\\Generic",
         "LocationInformation":"-"
      }
   }
}
```

The `DeviceId` field could be splitted into several fields, showing information about the serial number, the product and the vendor, for example.

To do this, the next steps could be followed:

- The Eventchannel module sends the event to `analysisd`, where it is decoded in JSON format, containing the `DeviceId` field with all the mentioned fields in it.
- This JSON event will match a silent rule filtering by the Security channel and event ID 6416, which is the Windows event for a new external device being connected.
- The rule will pass the event to `integratord`, where a script will split all the fields contained in `DeviceId` and remove every other field that is not of an interest.
- Now splitted, it can either be written to a localfile monitored by `logcollector` or sent directly to `analysisd`, avoiding the process of running `logcollector` again.
- It will match the JSON decoder and the new fields would be separated.

This way it would be possible to build new rules matching by `usb.serial_number`, `usb.product`, `usb.vendor`, etc., as the user mentioned.

Kind regards,
Cristina

**Comment by MHichamB** [comment]
Hi all,

Do you have any update about this issue? Is it closed? I'm facing the same problem so i wanted to know if you have another solution..

Thank you very much

**Comment by cristgl** [comment]
Hi @MHichamB,

The issue is not closed, it is in our roadmap and will be developed as soon as possible. I am sorry for the inconveniences.

Regards,
Cristina

**Comment by LFBernardo** [comment]
This may have a larger use case as there are numerous applications that nest their data in fields that are required for quick identification (username, ip, workstation etc)

**Comment by Tukaram1992** [comment]
Is this issue resolved.

**Comment by Tukaram1992** [comment]
Is this issue resolved.

**Comment by Caba44** [comment]
Is this issue resolved???

**Comment by d-steffan** [comment]
Another use case would be the auditing of LDAP queries (EventID 1644). There the data is decoded as an array, so its ending up in wazuh as:

`data.win.eventdata.data: DC=lab,DC=local,  ( &amp;  (sAMAccountType=805306368)  (servicePrincipalName=*) ( !  (sAMAccountName=krbtgt) ) ( !  (userAccountControl&amp;2) ) ), 9, 2, 10.0.0.16:55058, subtree, [all], idx_sAMAccountType:9:N;, 179, 0, 0, 0, 0, 0, none, LAB\\Administrator`

This sample was to detect SPN lookups as done by kerberoasting attacks. Apart from all the data effectively being held in one field, the first data point (Client) is completely missing from data.win.eventdata.data and can only be found in the full-text representation in data.win.system.message:

```
Internal event: A client issued a search operation with the following options. 
 
Client:
10.0.0.16:55058 
Starting node:
DC=lab,DC=local 
Filter:
 ( &  (sAMAccountType=805306368)  (servicePrincipalName=*) ( !  (sAMAccountName=krbtgt) ) ( !  (userAccountControl&2) ) )  
Search scope:
subtree 
Attribute selection:
[all] 
Server controls:
 
Visited entries:
9 
Returned entries:
2 
Used indexes:
idx_sAMAccountType:9:N; 
Pages referenced:
179 
Pages read from disk:
0 
Pages preread from disk:
0 
Clean pages modified:
0 
Dirty pages modified:
0 
Search time (ms):
0 
Attributes Preventing Optimization:
none 
User:
LAB\Administrator
```

So it would be very very nice to have the option to jam a custom decoder inbetween to correctly parse this mess to have the proper values to corelate this types of logs with kerberos TGS issues.

**Comment by d-steffan** [comment]
Surely the problem is how MIcrosoft saves this data in the event, because there are no specific parameter names for the given data. Everything is named data:

```
<Event xmlns="http://schemas.microsoft.com/win/2004/08/events/event">
  <System>
    <Provider Name="Microsoft-Windows-ActiveDirectory_DomainService" Guid="{0e8478c5-3605-4e8c-8497-1e730c959516}" EventSourceName="NTDS General" />
    <EventID Qualifiers="16384">1644</EventID>
    <Version>0</Version>
    <Level>4</Level>
    <Task>15</Task>
    <Opcode>0</Opcode>
    <Keywords>0x8080000000000000</Keywords>
    <TimeCreated SystemTime="2024-08-20T10:38:45.5122018Z" />
    <EventRecordID>118539</EventRecordID>
    <Correlation />
    <Execution ProcessID="656" ThreadID="776" />
    <Channel>Directory Service</Channel>
    <Computer>dc.lab.local</Computer>
    <Security UserID="S-1-5-7" />
  </System>
  <EventData>
    <Data>RootDSE</Data>
    <Data> ( &  (sAMAccountType=805306368)  (servicePrincipalName=*) ( !  (sAMAccountName=krbtgt) ) ( !  (userAccountControl&2) ) )   </Data>
    <Data>2</Data>
    <Data>0</Data>
    <Data>SAM</Data>
    <Data>subtree</Data>
    <Data>[]</Data>
    <Data>
    </Data>
    <Data>idx_distinguishedName:1:N;</Data>
    <Data>12</Data>
    <Data>0</Data>
    <Data>0</Data>
    <Data>0</Data>
    <Data>0</Data>
    <Data>0</Data>
    <Data>none</Data>
    <Data>UNAVAILABLE</Data>
  </EventData>
</Event>
```

**Comment by Caba44** [comment]
Is it possible to connect a sibling to extract the necessary data from EventData yourself?

**Comment by d-steffan** [comment]
> Is it possible to connect a sibling to extract the necessary data from EventData yourself?

I don't think so, because you cannot connect another decoder, so you would have to match the data in the rules. You can use regex there but the matched data won't be saved as parameter (e.g. data.win.eventdata.targetUserName, or srcip) so any correlation won't be possible. I at least can't think of a way to do so.

**Comment by Caba44** [comment]
In the Rules Syntax description you can specify variables for the description field:
<description>Source port $(srcport) is detected.</description>

  <rule id="160012" level="4">
    <if_sid>61070</if_sid>
    <field name="win.system.providerName">^MSSQLSERVER$</field>
    <field name="win.system.eventID">33205</field>
   <field name="win.system.message" type="pcre2">(?s)^.*server_principal_name:**(\w+)**.*$</field>
    <description>MS SQL **$(\1)**.</description>
  </rule>

**$(\1)** - empty :(

is it possible to somehow set a variable in a rule?

**Comment by Zer0Risk** [comment]
We would also need the possibility to add child decoders to "windows_eventchannel" decoded events.
Or some other way to edit events that got decoded by the "windows_eventchannel" decoder.
